### PR TITLE
Add strictness check and made tensors into leaves if input tensors were leaves

### PR DESCRIPTION
**Pull request by Chillee**
I think this makes sense to do? Otherwise, if you call `backward()` in your traced function, you can't get gradients out of any tensors that should have been leaves.

### CODE REVIEW

File: test/test_fx_experimental.py [hunk 1]
```diff
@@ -702,13 +702,18 @@ def forward(self, x):
...
+        def f_grad(x):
             val = x.cos().cos().sum()
             return torch.autograd.grad(val, x)
 
-        traced_graph = make_fx(f)(torch.randn(3, requires_grad=True))
-        inp = torch.randn(3, requires_grad=True)
-        torch.testing.assert_close(traced_graph(inp), f(inp))
+        def f_backward(x):
+            val = x.cos().cos().sum()
+            return torch.autograd.grad(val, x)
```
Comment (ezyang) [inline test/test_fx_experimental.py:711]: Did you mean backward here?

Comment (Chillee) [reply test/test_fx_experimental.py:711]: oops yeah

**Comment by ezyang** [comment]
If the input tensor isn't a leaf don't you still want to propagate gradients to it?

**Comment by Chillee** [comment]
@ezyang Well, in the current implementation, we never propagate inputs to a ProxyTensor, I believe.

But no, I don't think we should worry about propagating gradients to non-leaf tensors. In that case, `make_fx`'s semantics are just ill-defined imo.

**Comment by ezyang** [comment]
Thinking about this situation some more, I think all of the input `requires_grads` need to be "leaves" of some sort. But it doesn't matter for dynamo (since backward() will cause a graph break, right?)

Let me explain my reasoning. The problematic situation to consider is something like this:

```
def f(x):
  x.backward()

a = torch.randn(1, requires_grad=True)
b = a.sin()
aot_f = aot_function(f)
```

Under the new proposed semantics here, x would not be turned into a leaf inside AOTAutograd tracing, because it is not a leaf on entry. But to exactly replicate the semantics of the code in `aot_f`, we're obligated to mutate `a.grad` with some more gradients from the imperative backwards call whenever we call it. While hypothetically we can make this happen, it probably doesn't matter too much. But we should(!) raise an error if gradients try to escape!

File: torch/fx/experimental/proxy_tensor.py [hunk 2]
```diff
@@ -38,23 +38,34 @@ def decompose(decomposition_table):
...
             proxy.node.meta['tensor_meta'] = {}
-            r = torch.Tensor._make_subclass(cls, elem, elem.requires_grad)
         else:
-            r = super().__new__(cls, elem)  # type: ignore[call-arg]
             proxy.node.meta['tensor_meta'] = _extract_tensor_metadata(r)
         r.proxy = proxy  # type: ignore[attr-defined]
 
         return r
 
+    def __deepcopy__(self, memo):
```
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/fx/experimental/proxy_tensor.py:66]: With CUDA tensors, it errors with 

```
  File "/scratch/chilli/fresh/pytorch/torch/_tensor.py", line 149, in __deepcopy__
    new_tensor.set_(new_storage, self.storage_offset(), self.size(), self.stride())
  File "/scratch/chilli/fresh/functorch/functorch/_src/python_key.py", line 91, in __torch_dispatch__
    proxy_out = func(*proxy_args, **proxy_kwargs)
  File "/scratch/chilli/fresh/pytorch/torch/_ops.py", line 143, in __call__
    return self._op(*args, **kwargs or {})
  File "/scratch/chilli/fresh/pytorch/torch/fx/proxy.py", line 319, in __torch_function__
    return tracer.create_proxy('call_function', orig_method, args, kwargs,
  File "/scratch/chilli/fresh/pytorch/torch/fx/proxy.py", line 63, in create_proxy
    args_ = self.create_arg(args)
  File "/scratch/chilli/fresh/functorch/functorch/_src/python_key.py", line 162, in create_arg
    return super().create_arg(a)
  File "/scratch/chilli/fresh/pytorch/torch/fx/_symbolic_trace.py", line 291, in create_arg
    return super().create_arg(a)
  File "/scratch/chilli/fresh/pytorch/torch/fx/proxy.py", line 125, in create_arg
    return type(a)(self.create_arg(elem) for elem in a)
  File "/scratch/chilli/fresh/pytorch/torch/fx/proxy.py", line 125, in <genexpr>
    return type(a)(self.create_arg(elem) for elem in a)
  File "/scratch/chilli/fresh/functorch/functorch/_src/python_key.py", line 162, in create_arg
    return super().create_arg(a)
  File "/scratch/chilli/fresh/pytorch/torch/fx/_symbolic_trace.py", line 291, in create_arg
    return super().create_arg(a)
  File "/scratch/chilli/fresh/pytorch/torch/fx/proxy.py", line 151, in create_arg
    raise NotImplementedError(f"argument of type: {type(a)}")
NotImplementedError: argument of type: <class 'torch.cuda._UntypedStorage'>
```

[the remark this answers is not in the document]
Comment (ezyang) [reply torch/fx/experimental/proxy_tensor.py:66]: cc @kurtamohler

**Comment by Chillee** [comment]
@ezyang Oops, thought I already responded to this.

> But to exactly replicate the semantics of the code in aot_f, we're obligated to mutate a.grad with some more gradients from the imperative backwards call whenever we call it. While hypothetically we can make this happen, it probably doesn't matter too much. But we should(!) raise an error if gradients try to escape!

So... I think it's basically impossible to capture your case (and we also shouldn't even try). Currently, we never propagate gradients to the input (which is wrong). This PR simply fixes it for the "easy" case.

**Comment by Chillee** [comment]
@ezyang @albanD Could I get an approval on this PR? There are some ... tougher questions to answer that are unresolved, but I need this PR to replace functorch's tracer with this one.

**Review by ezyang — APPROVED** [review]

File: torch/fx/experimental/proxy_tensor.py [hunk 3]
```diff
@@ -172,7 +184,7 @@ def wrapped(*args):
         for idx, arg in enumerate(flat_args):
             if isinstance(flat_inps[idx], torch.Tensor):
                 with no_dispatch():
-                    flat_args[idx] = ProxyTensor(flat_inps[idx], arg)
+                    flat_args[idx] = ProxyTensor(flat_inps[idx], arg, requires_grad=flat_inps[idx].is_leaf)
```
Comment (byronyi) [inline torch/fx/experimental/proxy_tensor.py:187]: This errors out with tensor that is leaf but does not require gradient, e.g. int or boolean tensor.

```
RuntimeError: Only Tensors of floating point and complex dtype can require gradients
```

**Pull request closed without merging**